File: data-structures/queue/tests/queue.test.ts
import assert from "node:assert/strict";
import { MyQueue } from "../implementations/queue";

function runTest(name: string, execute: () => void): void {
  try {
    execute();
    console.log(`PASS ${name}`);
  } catch (error) {
    console.error(`FAIL ${name}`);
    throw error;
  }
}

function createReferenceSequence(): number[] {
  const values: number[] = [];

  for (let index = 1; index <= 50; index += 1) {
    values.push(index);
    if (index % 5 === 0) {
      values.shift();
    }
    if (index % 8 === 0) {
      values.push(index * 10);
    }
  }

  return values;
}

runTest("simple: new queue starts empty", () => {
  const queue = new MyQueue<number>(2);

  assert.equal(queue.length(), 0);
  assert.equal(queue.capacity(), 2);
  assert.equal(queue.isEmpty(), true);
  assert.equal(queue.peek(), undefined);
  assert.deepEqual(queue.toArray(), []);
});

runTest("simple: enqueue and peek expose the current front", () => {
  const queue = new MyQueue<number>(2);

  queue.enqueue(10);
  queue.enqueue(20);
  queue.enqueue(30);

  assert.equal(queue.length(), 3);
  assert.equal(queue.peek(), 10);
  assert.deepEqual(queue.toArray(), [10, 20, 30]);
});

runTest("simple: dequeue returns items in FIFO order", () => {
  const queue = new MyQueue<string>(2);

  queue.enqueue("a");
  queue.enqueue("b");
  queue.enqueue("c");

  assert.equal(queue.dequeue(), "a");
  assert.equal(queue.dequeue(), "b");
  assert.equal(queue.peek(), "c");
  assert.deepEqual(queue.toArray(), ["c"]);
});

runTest("simple: clear resets the queue", () => {
  const queue = new MyQueue<number>(2);

  queue.enqueue(1);
  queue.enqueue(2);
  queue.clear();

  assert.equal(queue.length(), 0);
  assert.equal(queue.isEmpty(), true);
  assert.equal(queue.dequeue(), undefined);
});

runTest("mildly absurd: queue grows beyond initial capacity", () => {
  const queue = new MyQueue<number>(1);

  for (let value = 0; value < 25; value += 1) {
    queue.enqueue(value);
  }

  assert.equal(queue.length(), 25);
  assert.ok(queue.capacity() >= 25);
  assert.equal(queue.peek(), 0);
});

runTest("mildly absurd: alternating enqueue and dequeue leaves the right survivors", () => {
  const queue = new MyQueue<number>(2);

  queue.enqueue(1);
  queue.enqueue(2);
  queue.dequeue();
  queue.enqueue(3);
  queue.enqueue(4);
  queue.dequeue();
  queue.enqueue(5);

  assert.deepEqual(queue.toArray(), [3, 4, 5]);
  assert.equal(queue.peek(), 3);
});

runTest("absurd: deterministic operation sequence matches a reference model", () => {
  const queue = new MyQueue<number>(2);
  const reference = createReferenceSequence();

  for (const value of reference) {
    queue.enqueue(value);
  }

  assert.deepEqual(queue.toArray(), reference);
  assert.equal(queue.length(), reference.length);
  assert.equal(queue.peek(), reference[0]);
});

File: data-structures/queue/tests/queue_test.go
package tests

import (
	"reflect"
	"testing"

	"dsa/data-structures/queue/implementations"
)

func createQueueReferenceSequence() []int {
	values := make([]int, 0)

	for index := 1; index <= 50; index++ {
		values = append(values, index)
		if index%5 == 0 {
			values = values[1:]
		}
		if index%8 == 0 {
			values = append(values, index*10)
		}
	}

	return values
}

func TestQueueSimpleNewQueueStartsEmpty(t *testing.T) {
	queue := implementations.NewMyQueue[int](2)

	if got := queue.Length(); got != 0 {
		t.Fatalf("expected length 0, got %d", got)
	}

	if got := queue.Capacity(); got != 2 {
		t.Fatalf("expected capacity 2, got %d", got)
	}

	if !queue.IsEmpty() {
		t.Fatal("expected queue to be empty")
	}

	if got := queue.ToSlice(); len(got) != 0 {
		t.Fatalf("expected empty slice, got %v", got)
	}
}

func TestQueueSimpleEnqueueAndPeekExposeCurrentFront(t *testing.T) {
	queue := implementations.NewMyQueue[int](2)

	queue.Enqueue(10)
	queue.Enqueue(20)
	queue.Enqueue(30)

	if got := queue.Length(); got != 3 {
		t.Fatalf("expected length 3, got %d", got)
	}

	if got := queue.Peek(); got != 10 {
		t.Fatalf("expected front value 10, got %d", got)
	}

	if got := queue.ToSlice(); !reflect.DeepEqual(got, []int{10, 20, 30}) {
		t.Fatalf("expected [10 20 30], got %v", got)
	}
}

func TestQueueSimpleDequeueReturnsItemsInFIFOOrder(t *testing.T) {
	queue := implementations.NewMyQueue[string](2)

	queue.Enqueue("a")
	queue.Enqueue("b")
	queue.Enqueue("c")

	if got := queue.Dequeue(); got != "a" {
		t.Fatalf("expected a, got %s", got)
	}

	if got := queue.Dequeue(); got != "b" {
		t.Fatalf("expected b, got %s", got)
	}

	if got := queue.Peek(); got != "c" {
		t.Fatalf("expected c, got %s", got)
	}
}

func TestQueueSimpleClearResetsTheQueue(t *testing.T) {
	queue := implementations.NewMyQueue[int](2)

	queue.Enqueue(1)
	queue.Enqueue(2)
	queue.Clear()

	if got := queue.Length(); got != 0 {
		t.Fatalf("expected length 0, got %d", got)
	}

	if !queue.IsEmpty() {
		t.Fatal("expected queue to be empty")
	}
}

func TestQueueMildlyAbsurdGrowthBeyondInitialCapacity(t *testing.T) {
	queue := implementations.NewMyQueue[int](1)

	for value := 0; value < 25; value++ {
		queue.Enqueue(value)
	}

	if got := queue.Length(); got != 25 {
		t.Fatalf("expected length 25, got %d", got)
	}

	if got := queue.Capacity(); got < 25 {
		t.Fatalf("expected capacity to grow beyond 25, got %d", got)
	}
}

func TestQueueMildlyAbsurdAlternatingEnqueueAndDequeueLeavesRightSurvivors(t *testing.T) {
	queue := implementations.NewMyQueue[int](2)

	queue.Enqueue(1)
	queue.Enqueue(2)
	queue.Dequeue()
	queue.Enqueue(3)
	queue.Enqueue(4)
	queue.Dequeue()
	queue.Enqueue(5)

	if got := queue.ToSlice(); !reflect.DeepEqual(got, []int{3, 4, 5}) {
		t.Fatalf("expected [3 4 5], got %v", got)
	}
}

func TestQueueAbsurdDeterministicOperationSequenceMatchesReferenceModel(t *testing.T) {
	queue := implementations.NewMyQueue[int](2)
	reference := createQueueReferenceSequence()

	for _, value := range reference {
		queue.Enqueue(value)
	}

	if got := queue.ToSlice(); !reflect.DeepEqual(got, reference) {
		t.Fatalf("expected %v, got %v", reference, got)
	}
}
